### DISCUSSION TITLE
Remove RTEMS 4.11 support in cFS?

### DISCUSSION THREAD

**Discussion by ghost** [nasa/cFS#467] [Polls]
We are working on adding RTEMS 6 support to the cFS. We currently support RTEMS 4.11 and RTEMS 5.
The purpose of this poll is to ask the community if RTEMS 4.11 support in the cFS is important to your future or current work.

I don't know any RTEMS 4.11 / cFS or OSAL Applications, so please chime in if you have an RTEMS 4.11 cFS  or OSAL Application you would like to see continued support for.

RTEMS 4.11 support includes the OSAL, PSP, Toolchain file, and CI workflow for RTEMS 4.11.

Thanks,
Alan

**Comment by jphickey** [comment]
The main thing that the RTEMS 4.11 build currently provides is a litmus test to check for issues with conversions between `unsigned int` and `unsigned long`.  Notably, the i686/pc-rtems build for 4.11 defines `uint32_t` as `unsigned long` -- whereas the significant majority of systems define it as `unsigned int` these days.  Therefore it becomes very easy to write code that assumes uint32 == unsigned int, which will fail to compile once the other type of system is encountered.

I did some experimentation and found that if I build a 32-bit SPARC "bare-metal" toolchain using the current version of crosstool-ng (1.25-rc2 as of this writing) that defines the int32 type as long, so it could be a feasible substitute.  There are probably others too.

I'd recommend identifying one and getting it into the CI matrix before retiring 4.11 completely.

**Reply by ghost** [reply]
I did an experiment with Gaisler's RCC toolchain (sparc-gaisler-rtems5-gcc , which is based on GCC 10.2.0) and it did not give me an error compiling this file:
```
#include <stdio.h>
void test_func(void)
{
   uint32_t val;
   val = 32;
   printf("%s(): val=%u\n",__func__, val);
}
```

Compile command is:
```
sparc-gaisler-rtems5-gcc -mcpu=leon3 -qrtems -qbsp=leon3 -Wall -Werror -o test.o -c test.c
```

Same results with "generic" sparc-rtems6 compiler, that is currently based on 10.3.1.

I do have instances where I get a warning with a print statement and the compiler recommends %lx or %ld or the other way around depending on the variable that I am printing. In this case, it is fine on Linux, but errors with a format warning on RTEMS. Is this the type of problem you are trying to check for?

**Reply by jphickey** [reply]
> I do have instances where I get a warning with a print statement and the compiler recommends %lx or %ld

In what cases do you need to add the %lu or %lx?  This is what we need to identify - a system that does in fact require the "l" designation here.  If that code compiled without error, that won't work.  We need to identify a system where it does require it the other way.

**Reply by ghost** [reply]
We use the Gaisler BCC (Bare C Compiler) GCC for diagnostic and boot code on the LEON platform. This is based on the old GCC 4.4.2. I had to add the stdint.h include to it to get it to compile:

```
#include <stdio.h>
#include <stdint.h>
void test_func(void)
{
   uint32_t val;
   val = 32;
   printf("%s(): val=%u\n",__func__, val);
}
```

But it does cause the warning:

```
cc1: warnings being treated as errors
test2.c: In function 'test_func':
test2.c:7: error: format '%u' expects type 'unsigned int', but argument 3 has type 'uint32_t'
```

I also noticed that the error occurs if I use the old common_types.h that typedef'ed a uint32 as unsigned long int instead of uint32_t. In this case, we are forcing the wrong type, correct?

The BCC compiler probably does not help, but I am starting to understand the problem better:
The goal is to make sure that we do not introduce any code into the cFS that mixes uint32_t and unsigned int in operations (assignments, etc). This passes on our newer systems, but fails on the older GCC based systems. If someone introduces this code, the RTEMS 4.11 build currently catches it.

Do you think a clang based compiler would catch it?
